File: javascript/leetcode/1019. Next Greater Node In Linked List.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {number[]}
 */
var nextLargerNodes = function (head) {
  const nums = [],
    stk = [];

  while (head) {
    nums.push(head.val);
    head = head.next;
  }

  for (let i = nums.length - 1; i >= 0; --i) {
    let val = nums[i];

    while (stk.length && stk[stk.length - 1] <= nums[i]) stk.pop();

    nums[i] = stk.length ? stk[stk.length - 1] : 0;
    stk.push(val);
  }

  return nums;
};

File: c++/leetcode/1019. Next Greater Node In Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
  vector<int> nextLargerNodes(ListNode *head)
  {
    vector<int> nums, stk;

    while (head)
    {
      nums.push_back(head->val);
      head = head->next;
    }

    for (int i = nums.size() - 1; i >= 0; --i)
    {
      auto val = nums[i];

      while (!stk.empty() && stk.back() <= nums[i])
        stk.pop_back();

      nums[i] = stk.empty() ? 0 : stk.back();
      stk.push_back(val);
    }

    return nums;
  }
};
